File: backend/src/main/java/com/lsoft/aguapotable/controller/MedidorController.java
package com.lsoft.aguapotable.controller;

import com.lsoft.aguapotable.domain.entity.Medidor;
import com.lsoft.aguapotable.domain.enums.EstadoMedidor;
import com.lsoft.aguapotable.dto.request.MedidorRequest;
import com.lsoft.aguapotable.service.MedidorService;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.security.SecurityRequirement;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/api/v1/medidores")
@RequiredArgsConstructor
@Tag(name = "Medidores", description = "Gestión de medidores de agua")
@SecurityRequirement(name = "bearerAuth")
public class MedidorController {

    private final MedidorService medidorService;

    @GetMapping("/suscriptor/{suscriptorId}")
    @Operation(summary = "Listar medidores de un suscriptor")
    public ResponseEntity<List<Medidor>> getPorSuscriptor(@PathVariable Long suscriptorId) {
        return ResponseEntity.ok(medidorService.getPorSuscriptor(suscriptorId));
    }

    @GetMapping
    @Operation(summary = "Listar medidores de la comunidad por estado")
    public ResponseEntity<List<Medidor>> listar(
            @RequestParam(required = false, defaultValue = "ACTIVO") EstadoMedidor estado,
            HttpServletRequest request) {
        Long comunidadId = (Long) request.getAttribute("comunidadId");
        return ResponseEntity.ok(medidorService.getPorComunidadYEstado(comunidadId, estado));
    }

    @GetMapping("/{id}")
    @Operation(summary = "Obtener medidor por ID")
    public ResponseEntity<Medidor> getById(@PathVariable Long id, HttpServletRequest request) {
        Long comunidadId = (Long) request.getAttribute("comunidadId");
        return ResponseEntity.ok(medidorService.obtenerPorId(id, comunidadId));
    }

    @PostMapping
    @PreAuthorize("hasAnyRole('SUPER_ADMIN','ADMIN','OPERADOR')")
    @Operation(summary = "Registrar nuevo medidor")
    public ResponseEntity<Medidor> crear(@Valid @RequestBody MedidorRequest dto,
                                          HttpServletRequest request) {
        Long comunidadId = (Long) request.getAttribute("comunidadId");
        return ResponseEntity.status(201).body(medidorService.crear(dto, comunidadId));
    }

    @PutMapping("/{id}")
    @PreAuthorize("hasAnyRole('SUPER_ADMIN','ADMIN','OPERADOR')")
    @Operation(summary = "Actualizar medidor")
    public ResponseEntity<Medidor> actualizar(@PathVariable Long id,
                                               @Valid @RequestBody MedidorRequest dto,
                                               HttpServletRequest request) {
        Long comunidadId = (Long) request.getAttribute("comunidadId");
        return ResponseEntity.ok(medidorService.actualizar(id, dto, comunidadId));
    }

    @PatchMapping("/{id}/baja")
    @PreAuthorize("hasAnyRole('SUPER_ADMIN','ADMIN')")
    @Operation(summary = "Dar de baja un medidor")
    public ResponseEntity<Medidor> darDeBaja(@PathVariable Long id, HttpServletRequest request) {
        Long comunidadId = (Long) request.getAttribute("comunidadId");
        return ResponseEntity.ok(medidorService.darDeBaja(id, comunidadId));
    }
}

File: backend/src/medidor/medidor.controller.ts
import {
  Controller,
  Get,
  Post,
  Put,
  Patch,
  Body,
  Param,
  Query,
  ParseIntPipe,
  HttpCode,
  UseGuards,
  DefaultValuePipe,
  ParseEnumPipe,
} from '@nestjs/common';
import { AuthGuard } from '@nestjs/passport';
import { ApiTags, ApiOperation, ApiBearerAuth } from '@nestjs/swagger';
import { MedidorService } from './medidor.service';
import { MedidorRequest } from './dto/medidor.dto';
import { EstadoMedidor } from '../common/enums';
import { ComunidadId } from '../common/decorators';
import { RolesGuard, Roles } from '../auth/guards/roles.guard';

@ApiTags('Medidores')
@ApiBearerAuth()
@Controller('api/v1/medidores')
@UseGuards(AuthGuard('jwt'), RolesGuard)
export class MedidorController {
  constructor(private readonly medidorService: MedidorService) {}

  @Get('suscriptor/:suscriptorId')
  @ApiOperation({ summary: 'Listar medidores de un suscriptor' })
  async getPorSuscriptor(
    @Param('suscriptorId', ParseIntPipe) suscriptorId: number,
  ) {
    return this.medidorService.getPorSuscriptor(suscriptorId);
  }

  @Get()
  @ApiOperation({ summary: 'Listar medidores de la comunidad por estado' })
  async listar(
    @ComunidadId() comunidadId: number,
    @Query('estado') estado?: string,
  ) {
    const estadoEnum =
      (estado as EstadoMedidor) || EstadoMedidor.ACTIVO;
    return this.medidorService.getPorComunidadYEstado(comunidadId, estadoEnum);
  }

  @Get(':id')
  @ApiOperation({ summary: 'Obtener medidor por ID' })
  async getById(
    @Param('id', ParseIntPipe) id: number,
    @ComunidadId() comunidadId: number,
  ) {
    return this.medidorService.obtenerPorId(id, comunidadId);
  }

  @Post()
  @Roles('SUPER_ADMIN', 'ADMIN', 'OPERADOR')
  @HttpCode(201)
  @ApiOperation({ summary: 'Registrar nuevo medidor' })
  async crear(
    @Body() dto: MedidorRequest,
    @ComunidadId() comunidadId: number,
  ) {
    return this.medidorService.crear(dto, comunidadId);
  }

  @Put(':id')
  @Roles('SUPER_ADMIN', 'ADMIN', 'OPERADOR')
  @ApiOperation({ summary: 'Actualizar medidor' })
  async actualizar(
    @Param('id', ParseIntPipe) id: number,
    @Body() dto: MedidorRequest,
    @ComunidadId() comunidadId: number,
  ) {
    return this.medidorService.actualizar(id, dto, comunidadId);
  }

  @Patch(':id/baja')
  @Roles('SUPER_ADMIN', 'ADMIN')
  @ApiOperation({ summary: 'Dar de baja un medidor' })
  async darDeBaja(
    @Param('id', ParseIntPipe) id: number,
    @ComunidadId() comunidadId: number,
  ) {
    return this.medidorService.darDeBaja(id, comunidadId);
  }
}
